[2단계 - 웹 기반 로또 게임] 데이지(김유경) 미션 제출합니다.

kimyou1102 opened the pull request:
## 학습 목표
이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.
- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트  
<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->
- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://kimyou1102.github.io/javascript-lotto/
 
## 구조 설계

각 관심사에 해당하는 컴포넌트들이 있고 상태 변경에 따라 동적으로 변화하는 부분이 재렌더되도록 설계했습니다. 상태는 전역으로 관리하고 `state`가 변경되면 구매폼을 제외한 부분에**서 값에 따라 나타나거나 사라지는 동적인 UI 요소**들이 다시 렌더링되는 방식으로 구현했습니다.

전역 상태 관리를 위해 Zustand의 패턴을 참고하여 Store를 만들고 아래와 같이 상태를 개별 Store 파일에서 관리하도록 설계했습니다.

- `lottoTransactionStore` : 구매한 금액과 발행된 로또 정보를 관리하는 Store
- `winningLottoInfoStore` : 당첨 번호 및 보너스 번호 정보를 관리하는 Store

## 리뷰 요청 & 논의하고 싶은 내용  
<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.  
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제
<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? --> 

전역 상태 관리를 도입한 이유는 **클래스를 사용할 때 인스턴스 변수를 3개 미만으로 제한하고, 함수의 매개변수를 2개 이하로 유지해야 한다는 요구사항** 때문입니다. 기존 방식에서는 상위 컴포넌트인 `LottoGame`에서 상태를 정의하고 이를 하위 컴포넌트로 전달하는 구조였기 때문에 매개변수와 인스턴스 변수가 많아지는 문제가 있었습니다.

이를 해결하기 위해 객체로 묶는 방법도 고려했지만 모든 상태를 객체로 묶을 것은 요구사항의 취지와 맞지 않다고 판단해서 전역 상태 관리 방식으로 변경하는 것이 더 적절하다고 생각했습니다!

하지만 이렇게 만들고 보니 props가 여러 번 전달되는 구조도 아니었는데 단순히 요구사항을 맞추기 위해 과도한 전역 상태 관리를 도입한 것은 아닌가 하는 고민이 남아 있습니다. 이에 대해 초파의 의견이 궁금합니다!

### 2) 이번 리뷰를 통해 논의하고 싶은 부분  
<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

1. **전역 상태 초기화 시 불필요한 렌더링을 줄이는 방법**

현재 전역 상태를 초기화할 때마다 각 Store의 `resetState()`가 호출되면서 구독한 모든 컴포넌트가 변경을 감지하여 여러 번 렌더링되는 문제가 발생합니다. 

상태를 초기화하는 과정에서 렌더링을 한 번만 발생하도록 최적화할 수 있는 방법이 있을지 고민해봤지만 명확한 해결 방법을 찾지 못했습니다.  리팩토링시 반영하고 싶어서 이에 대해서 쵸파가 힌트를 주신다면 감사하겠습니다!

2. **상태 변경 시 전체 렌더링 vs 부분 렌더링**

현재 상태가 변경될 때, 화면 전체가 다시 렌더링되도록 구현되어 있습니다.

이를 개선하기 위해 구독 범위를 더 세분화하여 특정 컴포넌트만 다시 렌더링하도록 변경하면 렌더링 비용을 줄일 수 있을 것 같지만,이 방식은 명세가 바뀔 경우 `subscribe()`에 등록된 코드가 계속 수정될 가능성이 있어 유지보수성이 떨어질 수 있다 생각했습니다.

렌더링 비용 절감과 유지보수성을 동시에 고려하면서 최적화할 수 있는 방법이 있을지 쵸파의 의견을 듣고 싶습니다!

3. **사용자 경험 개선**

모달, 인풋 등을 사용자가 편리하게 신경쓸 수 있도록 신경쓰며 기능들을 추가했습니다. 제가 한 기능들 외에도 개선할 사항들이 있는지 궁금합니다!!

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계
- [ ] TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [ ] 도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [ ] 객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [ ] 불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

### 2단계 
- [ ] 도메인 로직에 불필요한 영향을 주지 않고 UI 변경에 대응했는가?
- [ ] DOM 조작과 이벤트 활용을 JavaScript의 개념에 맞게 이해하고 적절하게 적용했는가? 
- [ ] 웹 표준을 준수하는 마크업을 활용하며, 스타일 작성에 일관성이 있는가?
</details>

kimyou1102 committed:
docs: 기능 요구 사항 작성

kimyou1102 committed:
feat: Nav 컴포넌트 생성

- nav 컴포넌트에서 사용되는 css

kimyou1102 committed:
style: Nav 컴포넌트 스타일 작성

kimyou1102 committed:
docs: Nav컴포넌트 구현 체크박스 표시

kimyou1102 committed:
feat: LottoGame 컴포넌트 생성

kimyou1102 committed:
style: LottoGame 컴포넌트 스타일 추가

kimyou1102 committed:
feat: index.js에 App컴포넌트 선언 및 Nav, LottoGame 컴포넌트 호출

kimyou1102 committed:
docs: LottoGame 컴포넌트 구현 사항 체크표시

kimyou1102 committed:
feat: PurchaseForm 컴포넌트 생성

- LottoGame 컴포넌트에 선언

kimyou1102 committed:
refactor: LottoGame을 감싸는 태그를 매개변수로 받아오도록 수정

- 로또 게임박스에 해당하는 영역만 관리하도록 하기 위해 수정

kimyou1102 committed:
feat: LottoHistoryItem 컴포넌트 생성

kimyou1102 committed:
feat: LottoPurchaseHistory 컴포넌트 생성

kimyou1102 committed:
refactor: LottoGame에서 상태를 관리하도록 구조 변경

- 하위 컴포넌트에 상태 변경 메서드를 넘겨주게 변경

kimyou1102 committed:
feat: 구매 버튼 클릭 시 로또 목록 및 당첨 번호 및 보너스 번호 입력 폼 렌더링

kimyou1102 committed:
feat: 공통 Button 컴포넌트 생성

kimyou1102 committed:
feat: LottoWinningInfoForm 컴포넌트 생성

kimyou1102 committed:
refactor: 기존 클래스가 유지되도록 classList.add 방식으로 변경

kimyou1102 committed:
feat: 모달 컴포넌트 생성

- 닫히는 기능까지 구현

kimyou1102 committed:
feat: 결과 확인하기 버튼 클릭 시 모달 오픈 기능 구현

kimyou1102 committed:
refactor: 공통 Button컴포넌트에 type="button" 지정

kimyou1102 committed:
feat: LottoWinningInfo 컴포넌트 생성

kimyou1102 committed:
refactor: 모달이 자식 컴포넌트의 렌더 메서드를 실행하도록 변경

kimyou1102 committed:
refactor: show 상태를 만들어 상태값에 따라 컴포넌트를 보여주게 수정

kimyou1102 committed:
feat: WinningForm에서 입력한 당첨 번호 및 보너스 번호를 상위 컴포넌트 상태로 반영하도록 구현

kimyou1102 committed:
fix: render시마다 lottos 가 새로 발행되는 문제 해결

- LottoGame의 render에서 lottos 객체 생성 로직 제거
- PurchaseForm 컴포넌트의 구입 버튼 클릭시 생성하도록 변경

kimyou1102 committed:
refactor: WinningStatistic 클래스 생성자 매개변수로 lottoHistory, rate를 받아와서 렌더하도록 수정

kimyou1102 committed:
feat: 당첨 통계 컴포넌트를 선언 및 당첨 결과를 계산하여 생성자 매개변수로 전달하도록 구현

kimyou1102 committed:
feat: 재시작 기능 구현

- 속성을 초기화 하는 메서드 생성 - setInti
- "다시 시작하기" 버튼 클릭시 setInit 메서드 호출

kimyou1102 committed:
refactor: modal닫힘 버튼 클릭 시 modal-none 클래스 추가하도록 변경

kimyou1102 committed:
refactor: 사용하지 않는 코드 제거

kimyou1102 committed:
docs: 구현한 사항들 체크 표시

kimyou1102 committed:
docs: 구현 사항 추가 작성

kimyou1102 committed:
feat: 구입할 금액에 대해 잘못된 입력 시 alert 로 에러 메시지 표시

kimyou1102 committed:
feat: 당첨 번호 및 보너스 번호 입력에 대해 잘못된 입력 시 alert로 에러 메시지 표시

kimyou1102 committed:
feat: Footer 컴포넌트 생성 및 선언

kimyou1102 committed:
refactor: Nav 컴포넌트의 생성자에서 append할 요소를 매개변수로 받도록 변경

kimyou1102 committed:
refactor: common폴더 components폴더 아래에 위치하도록 변경

- 공통 컴포넌트가 담긴 common폴더 components폴더 아래에 위치하도록 변경

kimyou1102 committed:
style: 당첨 확인 컨테이너 컴포넌트에 overflow-y: auto 적용하여 내용이 많을 때 스크롤 가능하도록 수정

kimyou1102 committed:
chore: 불필요한 주석 제거

kimyou1102 committed:
refactor: 컴포넌트별 CSS 파일을 분리하여 관리하도록 변경

kimyou1102 committed:
refactor: 공통 검증 메서드를 객체로 분리하고 필요한 검증만 선택적으로 사용하도록 변경

kimyou1102 committed:
test: 변경된 검증 로직에 맞춰 테스트 변경

- 클래스가 아닌 함수에서 검증을 하는 경우 파일명 변경 - validateBonusNumber.js, validatePrice.js
- 로또 클래스에서 검증 하도록 한 변경에 맞춰 Lotto.test.js로 검증 테스트 코드 이동

kimyou1102 committed:
refactor: 사용하지 않는 파일 제거

- 변경된 검증 로직에서 더이상 사용되지 않는 테스트 및 클래스 제거

kimyou1102 committed:
test: 변경된 Validator 에 맞게 테스트 변경

kimyou1102 committed:
refactor: step1-index.js의 검증 변경된 검증 로직으로 변경

kimyou1102 committed:
refactor: "구입" 버튼 클릭 이벤트 핸들러에서 콜백 함수 분리

- handleSubmit 메서드를 생성

kimyou1102 committed:
feat: 구매 금액 입력 폼에서 엔터 입력 시 구입 버튼과 동일하게 동작하도록 기능 추가

kimyou1102 committed:
refactor: 불필요한 클래스 제거문 제거

- 해당 클래스를 추가하는 부분이 없어서 제거

kimyou1102 committed:
refactor: 로또 당첨 결과 계산 로직을 별도 메서드로 분리하여 가독성 개선

- calculateWinningResult 메서드를 생성하여 분리

kimyou1102 committed:
refactor: 상수화된 값을 활용하도록 코드 수정

kimyou1102 committed:
refactor: "결과 확인하기" 버튼 클릭 이벤트 핸들러에서 콜백 함수 분리

- handleResultButtonClick 메서드 생성

kimyou1102 committed:
feat: 당첨 번호 및 보너스 번호 입력값이 초기화되지 않고 유지되도록 개선

kimyou1102 committed:
refactor: 불필요한 매개변수 제거 및 함수 내부에서 countNumber 구하도록 변경

kimyou1102 committed:
refactor: 매개변수 개수를 줄이기 위해 객체로 전달하도록 변경

kimyou1102 committed:
feat: 전역 상태 관리를 위한 store 파일 생성

kimyou1102 committed:
feat: lottoTransactionStore 생성하여 구입 내역 상태 관리 추가

kimyou1102 committed:
refactor: lottoTransaction을 lottoTransactionStore로 전역 상태 관리하도록 변경

kimyou1102 committed:
feat: winningLottoInfoStore 생성하여 당첨 로또 정보 상태 관리 추가

kimyou1102 committed:
refactor: winningLottoInfo을 winningLottoInfoStore로 전역 상태 관리하도록 변경

kimyou1102 committed:
refactor: LottoHistoryItem 컴포넌트 innerHTML로 작성된 부분 변경

- XSS 공격에 취약한 innerHTML대신 요소를 직접 생성해서 추가하는 식으로 변경

kimyou1102 committed:
feat: WinningNumbersInput 컴포넌트 생성

kimyou1102 committed:
feat: BonusNumberInput 컴포넌트 생성

kimyou1102 committed:
feat: 태그 생성을 위한 customCreateElement 유틸함수 생성

- tagName, className, text 속성을 가진 객체를 매개변수로 받아서 생성하는 유틸함수

kimyou1102 committed:
refactor: LottoWinningInfoForm에서 innerHTML 제거 및 입력 컴포넌트 분리

- LottoWinningInfoForm에서 innerHTML 사용 제거
- 보너스 번호 입력을 BonusNumberInput 컴포넌트로 분리
- 당첨 번호 입력을 WinningNumbersInput 컴포넌트로 분리

kimyou1102 committed:
feat: StatisticsTable 컴포넌트 생성

kimyou1102 committed:
refactor: 통계테이블을 StatisticsTable 컴포넌트로 변경

kimyou1102 committed:
style: StatisticsTable 스타일을 분리하여 독립적인 CSS 파일로 관리

kimyou1102 committed:
style: 공통 버튼 스타일에 cursor: pointer 스타일 추가

kimyou1102 committed:
fix tableHead 요소를 잘못된 순서로 선언했던 것을 수정

kimyou1102 committed:
refactor: 직접 요소 만드는 부분 customCreateElement를 사용하도록 변경

kimyou1102 committed:
refactor: Modal 닫힘 UX 개선

- 모달 배경 클릭 시 닫히는 기능 추가
- ESC눌렀을 시 닫히는 기능 추가

kimyou1102 committed:
feat: 구입 금액 입력 후 첫 번째 당첨 번호 입력칸에 자동 포커스 이동

kimyou1102 committed:
fix: Validator.range 사용시 누락된 매개변수 추가

kimyou1102 committed:
feat: 당첨 번호 입력 시 엔터로도 결과 확인 가능하도록 기능 추가

kimyou1102 committed:
refactor: 원본 배열을 변경하지 않도록 reverse를 toReversed로 변경

kimyou1102 committed:
fix: validate 함수의 빈 값 검사가 정상 동작하도록 Number 변환 과정 제거

kimyou1102 committed:
style: 제목 요소에 margin-bottom 추가를 위해 margin-top을 margin 변경

kimyou1102 committed:
refactor: 로또 번호 입력 그룹을 fieldset과 legend로 감싸 접근성 향상

kimyou1102 committed:
refactor: span 대신 label 사용하여 입력 필드 접근성 개선

kimyou1102 committed:
refactor: 원래 있던 style을 사용하기 위해 tag명만 바꿔서 생성하도록 변경

kimyou1102 committed:
fix: p 태그 대신 form이 생성되는 문제 수정

kimyou1102 committed:
refactor: 전체 박스가 아닌 로또 리스트에만 스크롤 적용하여 인풋이 내려가는 문제 수정

kimyou1102 committed:
fix: footer가 화면 축소 시 정상적으로 위치하도록 수정

kimyou1102 committed:
refactor: customCreateElement 유틸함수로 요소 만드는 것으로 변경

- innerHTML 제거

kimyou1102 committed:
refactor: innerHTML 대신 replaceChildren 사용하여 요소 초기화 방식 개선

kimyou1102 committed:
refactor: 정적인 요소는 한번만 렌더되도록 분리

- 정적인 요소만 렌더하는 renderStaticElement 메서드 생성

kimyou1102 committed:
refactor: render 될때 구매 입력 input value를 초기화 시키게 수정

kimyou1102 committed:
refactor: 정적 렌더 요소 초기화 로직을 메서드로 분리

kimyou1102 committed:
refactor: 사용하지 않는 코드 삭제

kimyou1102 committed:
refactor: store에 state 초기화 하는 함수 생성 및 적용

kimyou1102 committed:
refactor: 버튼 클릭 이벤트 제거하고 form의 submit 이벤트로 통합

kimyou1102 committed:
refactor: 공통 Button 컴포넌트에서 type을 매개변수로 설정할 수 있도록 개선

kimyou1102 committed:
refactor: 버튼 클릭 이벤트 제거하고 form의 submit 이벤트로 통합

kimyou1102 committed:
refactor: 불필요한 form태그 div태그로 변경

kimyou1102 committed:
refactor: 스타일 유지 목적의 div 래퍼 추가

kimyou1102 committed:
chore: homepage 값에 나의 GitHub username으로 변경

kimyou1102 committed:
fix: 배포 시 아이콘 이미지 경로 오류 수정

kimyou1102 committed:
fix: 배포 시 아이콘 이미지 경로 오류 수정

bassyu commented on src/components/BonusNumberInput.js:
```diff
@@ -0,0 +1,25 @@
+import customCreateElement from "../utils/customElement.js";
+
+export default class BonusNumberInput {
+  constructor($target, bonusNumber) {
+    this.render($target, bonusNumber);
+  }
+
+  render($target, bonusNumber) {
+    const $div = customCreateElement({
+      tagName: "div",
+      className: "lotto-number-input-wrap bonus-input-wrap",
+    });
+
+    const $text = customCreateElement({ tagName: "span", text: "보너스 번호" });
+    const $input = customCreateElement({
+      tagName: "input",
+      className: "number-input",
+    });
+    $input.value = bonusNumber === 0 ? "" : bonusNumber;
+
+    $div.appendChild($text);
+    $div.appendChild($input);
+    $target.appendChild($div);
+  }
+}
```
이렇게도 가능할 것 같은데요~ 클래스로 작성해주신 이유가 궁금합니다~!

```suggestion
export default function renderBonusNumberInput($target, bonusNumber) {
  const $div = customCreateElement({
    tagName: "div",
    className: "lotto-number-input-wrap bonus-input-wrap",
  });

  const $text = customCreateElement({ tagName: "span", text: "보너스 번호" });
  const $input = customCreateElement({
    tagName: "input",
    className: "number-input",
  });
  $input.value = bonusNumber === 0 ? "" : bonusNumber;

  $div.appendChild($text);
  $div.appendChild($input);
  $target.appendChild($div);
}
```

kimyou1102 replied:
처음에 컴포넌트의 상태를 클래스 속성처럼 관리하기 위해 클래스를 사용했습니다.
상태 변경은 setState와 같은 메서드를 통해서만 이루어지도록 하고
다른 메서드에서는 공통적으로 해당 속성을 참조할 수 있도록 클래스 구조를 선택했습니다. 그래서 전체적으로 컴포넌트는 다 클래스 구조를 갖도록 했습니다!

bassyu replied:
거의 모든 클래스에 setState가 없으니, 함수가 좀 더 어울린다고 생각했어요~!
예전에 setState가 있었던 레거시라고 이해하면 될까요?

kimyou1102 replied:
넵! 그리고 render를 따로 해주는 부분이 있어서 엘리먼트 생성부분과 render부분을 나눌 수 있도록 클래스를 사용한 이유도 있습니다!
통일성도 좋지만 이런 단순한 부분은 어쩌면 함수가 더 어울리는게 아닌가 하는 고민도 드는 거 같습니다

bassyu commented on src/components/LottoGame.js:
```diff
@@ -0,0 +1,113 @@
+
+    const winningStatistic = new WinningStatistic(winningResult, this.setInit);
+
+    new Modal(this.#target, ($target) => winningStatistic.render($target));
```
모든 컴포넌트에서 target 파라미터를 직접 받는 이유가 궁금합니다~!
appendChild만 하는 것이 목적이라면, 요소를 반환받아서 직접 하는 것은 어떤가요?

```suggestion
    this.#target.appendChild(new Modal(winningStatistic.render()));
```

kimyou1102 replied:
$target을 넘겨주는 이유는 해당 컴포넌트를 렌더하고 싶은 곳이 달라질 수 있으니 확장성을 고려해서 $target을 받도록 했습니다. 그런데 이렇게 하면 붙여주는 곳을 알지 않아도 되어서 오히려 좋다고 생각됩니다! render를 "화면에 그린다" 는 생각만 해서 요소를 반환할 생각을 못했네요ㅎㅎ

bassyu commented on src/store/winningLottoInfo.js:
```diff
@@ -0,0 +1,14 @@
+    winningNumbers: [],
+    bonusNumber: 0,
+  },
+  setLottoTransaction: (newWinningLottoInfo) =>
```
사용되지 않고 있긴 하지만, 오타인 것 같습니다~!

bassyu commented on src/components/LottoGame.js:
components 폴더에 파일들이 플랫하게 있어서 DOM 트리의 위계를 파악하기 어렵네요!
비슷한 관심사끼리 묶거나, 폴더 구조로 UI의 위계를 그대로 표현해주는 방법은 어떤가요?
예를 들면 아래처럼 폴더를 더 깊게 만들어 볼 수 있습니다!
```
LottoGame
> LottoPurchaseHistory
>> LottoHistoryItem
```

추가로 css 파일도 컴포넌트와 1대1로 가깝게 위치시키는 것도 추천드립니다!

kimyou1102 replied:
보통 비슷한 관심사끼리 묶었어서 layout말고는 어떻게 할지 고민하다가 그대로 제출했는데 파악이 어려울 수 있겠네요. 파악하기 쉽도록 폴더 구조를 변경하여 반영하겠습니다!

css파일과 1대1로 가깝게 위치 시키는 것도 응집성이 높아져서 좋을 거 같습니다👍

bassyu commented on src/components/common/Modal.js:
```diff
@@ -0,0 +1,57 @@
+      }
+    });
+
+    document.addEventListener("keydown", (e) => {
```
여긴 이벤트 리스너가 계속 쌓여서 비효율이 있네요! 어떻게 해결할 수 있을까요?

kimyou1102 replied:
addEventListener를 하기 전에 기존에 이벤트 리스너를 removeEventListener로 제거해주면 될 거 같습니다!

kimyou1102 replied:
Modal이 `LottoGame`의 render에서 계속 새로 생겨나서 `removeEventListener`를 해줘도 새로 생긴 함수라서 등록된 것과 같은 것으로 인지를 못하더라구요. 그래서 keydown 이벤트 리스너 부분을 한번만 렌더되는 곳으로 이동하여 해결했습니다!

bassyu commented on src/validation/validateBonusNumber.js:
extends 대신 함수로 풀어주셨네요 👍 👍

bassyu reviewed (changes requested):
안녕하세요 데이지!
드디어 콘솔을 벗어나 DOM을 사용하게 되었네요! 😄 
2단계도 잘 부탁드립니다!

## 기능

잘 구현 해주셨네요!!
아래는 우선순위가 높다고 생각하는 요구사항입니다~!

1. 문자 전송 후 숫자만 넣어달라고 하기보다, 애초에 숫자만 입력할 수 있도록 컴포넌트를 수정하면 더 좋을 것 같습니다. 
2. 마찬가지로, 오직 범위 안의 숫자만 입력할 수 있으면 좋을 것 같아요! 범위가 어떻게 되는지는 꼭 알려주면 좋을 것 같습니다~!

만약 시간이 있다면, 

1. 전송하기 전에 오류 메세지를 미리 보여주면 좋을 것 같아요! 
예를 들어, 중복이 있으면 결과 보기 버튼 자체가 눌리지 않았으면 좋겠어요.
2. 키보드를 통해 당첨 번호를 이동하면 좋을 것 같습니다!

## 코드

UI 관심사를 잘 분리해서 구현해주셨네요! 👍 👍 

html이 텅 비어있고, dom api를 통해 모든 UI를 그려주셨는데요,
html로 작성하는 것과 비교해서 어떤 장단점이 있었는지 궁금합니다!

'상태' 개념도 많이 참고하신 것 같아요!
상태는 무엇이고, 만약 상태 개념을 사용하지 않았다면 어떻게 구현했을지, 그리고 상태가 없으면 어떤 장단점이 있을지 답변해주실 수 있을까요?


## 논의

> 하지만 이렇게 만들고 보니 props가 여러 번 전달되는 구조도 아니었는데 단순히 요구사항을 맞추기 위해 과도한 전역 상태 관리를 도입한 것은 아닌가 하는 고민이 남아 있습니다. 이에 대해 초파의 의견이 궁금합니다!

옵저버 패턴 구현 학습의 관점으로 보면 좋은 시도인 것 같습니다~!
(옵저버 패턴은 어떤 문제를 해결하기 위해 도입하셨나요~?)

다만 '전역 상태 관리의 도입 시기' 기준으로 봤을 땐 조금 이르다는 생각이 들었어요!
혹시 전역 상태가 왜 위험한지 알고 계신가요?

말씀해주신 함수의 파라미터 요구사항도 '어디서나 접근 가능한 데이터 사용'을 유도했다기 보다는, 함수의 역할을 작고 순수하게 나누기 위해서 라고 생각합니다~! 

> 전역 상태 초기화 시 불필요한 렌더링을 줄이는 방법

제가 봤을 때에도 지금 구현으로는 store가 분리되어 있어서 어쩔 수 없겠네요!
모든 구독을 알고있는 무언가가 render를 직접 호출해야 해결될 것 같습니다.

> 렌더링 비용 절감과 유지보수성을 동시에 고려하면서 최적화할 수 있는 방법이 있을지 쵸파의 의견을 듣고 싶습니다!

말씀주신 대로 세분화해서 데이터가 필요한 컴포넌트만 구독하면 해결될 것 같은데요,
어떤 명세가 들어왔을 때 어떻게 수정되는지 예시를 들어주실 수 있을까요~? 

---

많은 시도를 하며 구현해주셨네요~!
질문을 많이 드렸는데, 충분히 고민해보시고 답변주세요!
수고 많으셨습니다 데이지! 💪 💪

kimyou1102 commented:
안녕하세요 쵸파! 쵸파가 남겨주신 질문들과 리뷰를 보며 생각해보지 못한 점들을 고민해보는 시간을 가질 수 있어서 좋았습니다😊

> html이 텅 비어있고, dom api를 통해 모든 UI를 그려주셨는데요, html로 작성하는 것과 비교해서 어떤 장단점이 있었는지 궁금합니다!

- 장점: 동적으로 변하는 부분에 대해서 그려주는 것이 가능합니다. 그리고 `querySelector` 를 하지 않고 생성한 요소 변수를 직접 사용할 수 있습니다.
- 단점: 일일이 선언해야 해서 코드가 복잡해집니다. 또한 js로 렌더링을 한다면 HTML파일에 미리 작성된 경우보다 처리하는 데 시간이 더 걸릴 수 있습니다.


> ‘상태' 개념도 많이 참고하신 것 같아요! 상태는 무엇이고, 만약 상태 개념을 사용하지 않았다면 어떻게 구현했을지, 그리고 상태가 없으면 어떤 장단점이 있을지 답변해주실 수 있을까요?

- 상태는 컴포넌트를 이루는 변경될 수 있는 값이라 생각합니다. 이 값의 변화에 따라 화면이 달라지는 것들을 상태로 관리한다 생각했습니다!
- 사용하지 않았다면 DOM API를 이용해 값을 직접 가져와 연산을 했을 것 같습니다. 이벤트가 발생할 때마다  DOM API를 조작해 화면을 업데이트 했을 거 같습니다. 그리고 `lottos`처럼 연산한 값을 저장해야 하는 거라면 전역변수로 두고 해당 변수를 이용했을 거 같습니다.
- 상태가 없을 때 장점:
    - 상태가 변경되어서 일어나는 리렌더링이나 사이드 이펙트의 같은 것을 관리해야 하는 부담이 없습니다.
- 상태가 없을 때의 단점:
    - 지금과 같은 경우는 상태만 변경하면 UI가 알아서 갱신되는데 이것과 다르게 일일이 DOM API를 직접 조작하며 변경시키줘야 하는 것이 번거롭고 각 영역마다 바꿔줘야 하니 오류가 생길 가능성이 있습니다.
    - 각각의 변경사항을 따로 관리하다보면 유지보수가 어려워질 수 있을 거 같습니다.
    - UI가 변경되는 코드를 여러곳에서 중복적으로 작성해야 하는 경우도 생깁니다.

> 옵저버 패턴 구현 학습의 관점으로 보면 좋은 시도인 것 같습니다~!
(옵저버 패턴은 어떤 문제를 해결하기 위해 도입하셨나요~?) 

상태가 변경된 것을 구독한 컴포넌트가 자동으로 업데이트 되도록 하기 위해 도입했습니다. zustand를 모방하는 것에만 급급해서 그런지 옵저버 패턴을 제대로 활용하지는 못한 거 같아요😢 render 메서드를 세분화 해서 제대로 활용할 수 있도록 해보겠습니다!

> 혹시 전역 상태가 왜 위험한지 알고 계신가요? 

자세히는 알지 못해 찾아보니 컴포넌트가 직접 상태를 참조하고 수정할 수 있기 때문에 컴포넌트의 역할이 많아지고 복잡해질 위험이 있습니다. 또한 데이터를 제공하는 곳과 사용하는 곳이 명확히 연결되지 않아서 상태를 어디서 변경했는지 추적하기 어려운 문제가 발생할 수 있습니다. 그리고 여러곳에서 상태를 변경하는 것이 가능하다 보니 디버깅이 어려워질 수도 있습니다. 장단점을 비교하여 맞게 사용했어야 했는데 요구사항에 매몰되어 작성한 거 같습니다..…🙃

> 말씀주신 대로 세분화해서 데이터가 필요한 컴포넌트만 구독하면 해결될 것 같은데요, 어떤 명세가 들어왔을 때 어떻게 수정되는지 예시를 들어주실 수 있을까요~?

현재 `price` 값은 통계 UI에서만 사용되므로, 해당 UI만 구독하면 해결할 수 있습니다. 하지만, 만약 구조가 변경되어 `purchaseForm` 등 다른 UI에서도 값이 필요해진다면 통계 UI렌더뿐 아니라 구매폼 UI렌더도 실행해야합니다.

이 경우 `lottoTransactionStore`의 구독 로직이 변경되어야 하기 때문에 유지보수 범위가 커질 수 있다고 생각했습니다. 하지만 render를 해주는 메서드를 잘 세분화한다면 필요한 렌더 메서드만 가져와서 사용하면 되니 예상보다 간단하게 해결할 수 있을 거 같습니다!

kimyou1102 committed:
refactor: 메서드명 오타 수정

kimyou1102 committed:
refactor: setState 직접 호출을 제거하고 기존 상태 변경 메서드 활용

kimyou1102 committed:
refactor: 인풋 number만 입력 되도록 변경

kimyou1102 committed:
refactor: 입력값의 최소 최대값 지정

kimyou1102 committed:
refactor: Button 컴포넌트 render 메서드에서 요소 리턴하는 방식으로 변경

kimyou1102 committed:
refactor: 당첨 번호끼리 중복되는 경우 버튼 비활성화 처리

kimyou1102 committed:
refactor: 당첨 번호 중에 보너스번호와 중복되는 숫자가 있다면 결과 버튼 비활성화 처리

kimyou1102 committed:
refactor: 폴더 구조 개선

- 관심사별로 폴더를 만들어서 정리

kimyou1102 committed:
refactor: keydown 이벤트 리스너 중복 호출되는 문제 해결

- Modal컴포넌트가 재렌더마다 새로 그려지며 keydown을 중복 등록하는 문제 발생
- 한번만 렌더되는 부분으로 keydown이벤트 리스너를 이동하여 해결

bassyu reviewed (approved):
안녕하세요 데이지!
데이지의 고민을 다시 한번 정리하면 좋을 것 같아서 질문을 많이 드렸어요!
잘 고민 해주시고 정리해 주셔서 감사합니다! 👍 👍 

짧은 시간에도 기능을 우선적으로 반영해주셨네요! 
아주 좋은 포인트라고 생각합니다! 💯 💯 
기능을 우선적으로 반영해주시면서, 동시에 '변화에 유연한 코드는 무엇일지?' 고민해보는 것을 강력 추천드릴게요!

전역 데이터의 특징도 잘 설명주셨네요!
유지보수 관점에서 데이터의 추적이 용이한 것은, 꽤 중요한 포인트라고 생각합니다.
또한 전역 데이터는 값의 초기화를 따로 관리해줘야 하는데요, 이것이 장점일 수도 있지만, 버그로 변할 가능성이 높으니 특히 주의하면 좋을 것 같아요~!

수고 많으셨습니다 데이지!
앞으로 미션도 파이팅입니다! 💪 💪 💪
